libbeat/common/kubernetes/watcher.go

```diff
@@ -0,0 +1,263 @@
+	ListenUpdate() bus.Listener
+
+	// ListenStop returns a bus listener to receive pod stopped events, with a `pod` key holding it
+	ListenStop() bus.Listener
+}
+
+type podWatcher struct {
+	sync.RWMutex
+	client              Client
+	syncPeriod          time.Duration
+	cleanupTimeout      time.Duration
+	nodeFilter          k8s.Option
+	lastResourceVersion string
+	ctx                 context.Context
+	stop                context.CancelFunc
+	bus                 bus.Bus
+	pods                map[string]*Pod      // pod id -> Pod
+	deleted             map[string]time.Time // deleted annotations key -> last access time
+}
+
+// Client for Kubernetes interface
+type Client interface {
+	ListPods(ctx context.Context, namespace string, options ...k8s.Option) (*corev1.PodList, error)
+	WatchPods(ctx context.Context, namespace string, options ...k8s.Option) (*k8s.CoreV1PodWatcher, error)
+}
+
+// NewWatcher initializes the watcher client to provide a local state of
+// pods from the cluster (filtered to the given host)
+func NewWatcher(client Client, syncPeriod, cleanupTimeout time.Duration, host string) Watcher {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &podWatcher{
+		client:              client,
+		cleanupTimeout:      cleanupTimeout,
+		syncPeriod:          syncPeriod,
+		nodeFilter:          k8s.QueryParam("fieldSelector", "spec.nodeName="+host),
+		lastResourceVersion: "0",
+		ctx:                 ctx,
+		stop:                cancel,
+		pods:                make(map[string]*Pod),
+		deleted:             make(map[string]time.Time),
+		bus:                 bus.New("kubernetes"),
+	}
+}
+
+func (p *podWatcher) syncPods() error {
+	logp.Info("kubernetes: %s", "Performing a pod sync")
+	pods, err := p.client.ListPods(
+		p.ctx,
+		"",
+		p.nodeFilter,
+		k8s.ResourceVersion(p.lastResourceVersion))
+
+	if err != nil {
+		return err
+	}
+
+	p.Lock()
+	for _, apiPod := range pods.Items {
+		pod := GetPod(apiPod)
+		p.pods[pod.Metadata.UID] = pod
+	}
+	p.Unlock()
+
+	// Emit all start events (avoid blocking if the bus get's blocked)
+	go func() {
+		for _, pod := range p.pods {
+			p.bus.Publish(bus.Event{
+				"start": true,
+				"pod":   pod,
+			})
+		}
+	}()
+
+	// Store last version
+	p.lastResourceVersion = pods.Metadata.GetResourceVersion()
+
+	logp.Info("kubernetes: %s", "Pod sync done")
+	return nil
+}
+
+// Start watching pods
+func (p *podWatcher) Start() error {
+
+	// Make sure that events don't flow into the annotator before informer is fully set up
+	// Sync initial state:
+	synced := make(chan struct{})
+	go func() {
+		p.syncPods()
+		close(synced)
+	}()
+
+	select {
+	case <-time.After(p.syncPeriod):
+		p.Stop()
+		return errors.New("Timeout while doing initial Kubernetes pods sync")
+	case <-synced:
+		// Watch for new changes
+		go p.watch()
+		go p.cleanupWorker()
+		return nil
+	}
+}
+
+func (p *podWatcher) watch() {
```

Comment:
do you think this is generic enough that it can be used by any implementations other than autodiscover and add_kubernetes_metadata? Would we need a scenario where we would want to write a controller which behaves a bit differently that we might want to initialize it with a different func that handles add/update/delete a bit differently?